[RISCV] Initial codegen support for zvqdotq extension

preames opened the pull request:
This patch adds pattern matching for the basic usages of the dot product instructions introduced by the experimental zvqdotq extension.  It specifically only handles the case where the pattern is feeding a i32 sum reduction as we need reassociate the reduction tree to use these instructions.

The vecreduce_add (sext) and vecreduce_add (zext) cases are included mostly to exercise the VX matchers.  For the generic matching, we fail to match due to an order of combine issue which results in the bitcast being separated from the splat.

I chose to do this lowering as an early combine so as to avoid having to integrate the entire logic into the reduction lowering flow.  In particular, that would get a lot more complicated as we extend this to handle add-trees feeding the reductions.

preames committed:
[RISCV] Initial codegen support for zvqdotq extension

This patch adds pattern matching for the basic usages of the dot product
instructions introduced by the experimental zvqdotq extension.  It
specifically only handles the case where the pattern is feeding a
i32 sum reduction as we need reassociate the reduction tree to use
these instructions.

The vecreduce_add (sext) and vecreduce_add (zext) cases are included
mostly to exercise the VX matchers.  For the generic matching, we
fail to match due to an order of combine issue which results in the
bitcast being separated from the splat.

I chose to do this lowering as an early combine so as to avoid
having to integrate the entire logic into the reduction lowering
flow.  In particular, that would get a lot more complicated as
we extend this to handle add-trees feeding the reductions.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -18003,6 +18003,118 @@ static SDValue performBUILD_VECTORCombine(SDNode *N, SelectionDAG &DAG,
+    return SDValue();
+
+  SDLoc DL(N);
+  MVT VT = N->getSimpleValueType(0);
```
We probably need to use EVT here?

topperc commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-zvqdotq.ll:
```diff
@@ -181,29 +221,77 @@ entry:
+; NODOT-LABEL: reduce_of_sext:
+; NODOT:       # %bb.0: # %entry
+; NODOT-NEXT:    vsetivli zero, 16, e32, m4, ta, ma
+; NODOT-NEXT:    vsext.vf4 v12, v8
```
Not related to this patch, but is this a missed opportunity to use vsext.vf2+vwredsum?

preames replied:
Yep, agreed.  I was a bit surprised by the codegen as I swear I've seen us do that in workloads.

preames committed:
Address review comment

mshockwave commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -18003,6 +18003,118 @@ static SDValue performBUILD_VECTORCombine(SDNode *N, SelectionDAG &DAG,
+      !InVec.getValueType().getVectorElementCount().isKnownMultipleOf(4))
+    return SDValue();
+
+  // reduce (sext a) <--> reduce (mul zext a. zext 1)
+  // reduce (zext a) <--> reduce (mul sext a. sext 1)
```
should it be `reduce (zext a) <--> reduce (mul zext a. zext 1)`?

preames replied:
Yep, I swapped them.  Fixing.

mshockwave commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -18003,6 +18003,118 @@ static SDValue performBUILD_VECTORCombine(SDNode *N, SelectionDAG &DAG,
+      return SDValue();
+    Opc = RISCVISD::VQDOTSU_VL;
+  }
+  assert(Opc);
```
will this ever got triggered?

preames replied:
It shouldn't with the current code structure, but asserts exist to check assumptions?

mshockwave commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -18003,6 +18003,118 @@ static SDValue performBUILD_VECTORCombine(SDNode *N, SelectionDAG &DAG,
+  // mul (zext, sext) -> vqdotsu (swapped)
+  // TODO: Improve .vx handling - we end up with a sub-vector insert
+  // which confuses the splat pattern matching.  Also, match vqdotus.vx
+  if (InVec.getOpcode() != ISD::MUL)
```
just curious, what about left shifts?

preames replied:
Annoyingly complicated, possible future work.

The problem is that we have to expand the shift as a multiply by 2^N, and the range of shift amounts we can handle is very limited due to the input being an i8.

preames replied:
p.s. I don't have a real case which benefits from this, if you do, please share and I'll re-prioritize.

preames committed:
Address review comment

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoZvqdotq.td:
```diff
@@ -26,3 +26,34 @@ let Predicates = [HasStdExtZvqdotq] in {
+// TODO: Add pseudo and patterns for vqdotus.vx
+let Predicates = [HasStdExtZvqdotq], mayLoad = 0, mayStore = 0,
+    hasSideEffects = 0 in {
+  defm PseudoVQDOT : VPseudoVQDOT_VV_VX;
```
Please add a FIXME to set isCommutable for VQDOT and VQDOTU. MachineCSE can use it, but it probably needs custom code and of course a test.

topperc reviewed (approved):
LGTM other than comment about adding a FIXME for isCommutable.

preames committed:
Address review comment

preames committed:
Merge branch 'main' into pr-riscv-zvqdotq-initial-codegen

s-barannikov commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -18101,6 +18101,118 @@ static SDValue performBUILD_VECTORCombine(SDNode *N, SelectionDAG &DAG,
+  const unsigned Policy = RISCVVType::TAIL_AGNOSTIC | RISCVVType::MASK_AGNOSTIC;
+  SDValue PolicyOp = DAG.getTargetConstant(Policy, DL, Subtarget.getXLenVT());
+  SDValue LocalAccum = DAG.getNode(Opc, DL, ContainerVT,
+                                   {Op0, Op1, Passthru, Mask, VL, PolicyOp});
```
This fails node verification (#138381).
VDOT_VL node is declared with 5 operands in td file, but is created with 6 operands.

preames replied:
Good catch.  Will push a fix shortly.  (The last operand is simply being ignored.)

preames replied:
Fixed in 92d949

s-barannikov commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -6971,7 +6971,7 @@ static bool hasPassthruOp(unsigned Opcode) {
+      RISCVISD::LAST_VL_VECTOR_OP - RISCVISD::FIRST_VL_VECTOR_OP == 139 &&
       RISCVISD::LAST_STRICTFP_OPCODE - RISCVISD::FIRST_STRICTFP_OPCODE == 21 &&
       "adding target specific op should update this function");
   if (Opcode >= RISCVISD::ADD_VL && Opcode <= RISCVISD::VFMAX_VL)
```
Shouldn't this return true for the added nodes?

preames replied:
The routines hasPassthruOp and hasMaskOp are only used by lowerToScalableOp and lowerVPOp.  These new nodes don't go through either of those, and thus the result is a don't care.  Honestly, the code structure of these routines could use some cleanup, but it's not a problem with the patch.

s-barannikov replied:
> Honestly, the code structure of these routines could use some cleanup

Those will be removed by #138381. Just wanted to make sure the `HasPassthruOp` flag should be set on those nodes.
CC @lenary

preames commented:
The github interface for this is now really confusing.  This change was *not* reverted.   I'd considered doing so, but the web-interface got confused, so I switched to just pushing a fix.  I fixed forward the build failure in 5f7213e0.  

The build breakage was because I'd merged main to resolve a conflict, and then forgot to commit the final edits on the merge before pushing to the PR and merging via the web-interface.  These half terminal and half web-interface flows are extremely error prone.
